libtransmission/peer-mgr-wishlist.cc

```diff
@@ -0,0 +1,237 @@
+/*
+ * This file Copyright (C) 2021 Mnemosyne LLC
+ *
+ * It may be used under the GNU GPL versions 2 or 3
+ * or any future license endorsed by Mnemosyne LLC.
+ *
+ */
+
+#include <algorithm>
+#include <cstddef>
+#include <iostream>
+#include <iterator>
+#include <numeric>
+#include <utility>
+#include <vector>
+
+#define LIBTRANSMISSION_PEER_MODULE
+
+#include "transmission.h"
+#include "crypto-utils.h" // tr_rand_buffer()
+#include "peer-mgr-wishlist.h"
+
+namespace
+{
+
+struct Candidate
+{
+    tr_piece_index_t piece;
+    size_t n_blocks_missing;
+    tr_priority_t priority;
+    uint8_t salt;
+
+    Candidate(tr_piece_index_t piece_in, size_t missing_in, tr_priority_t priority_in, uint8_t salt_in)
+        : piece{ piece_in }
+        , n_blocks_missing{ missing_in }
+        , priority{ priority_in }
+        , salt{ salt_in }
+    {
+    }
+
+    int compare(Candidate const& that) const // <=>
+    {
+        // prefer pieces closer to completion
+        if (n_blocks_missing != that.n_blocks_missing)
+        {
+            return n_blocks_missing < that.n_blocks_missing ? -1 : 1;
+        }
+
+        // prefer higher priority
+        if (priority != that.priority)
+        {
+            return priority > that.priority ? -1 : 1;
+        }
+
+        if (salt != that.salt)
+        {
+            return salt < that.salt ? -1 : 1;
+        }
+
+        return 0;
+    }
+
+    bool operator<(Candidate const& that) const // less than
+    {
+        return compare(that) < 0;
+    }
+};
+
+std::vector<Candidate> getCandidates(Wishlist::PeerInfo const& peer_info)
+{
+    // count up the pieces that we still want
+    auto wanted_pieces = std::vector<std::pair<tr_piece_index_t, size_t>>{};
+    auto const n_pieces = peer_info.countAllPieces();
+    wanted_pieces.reserve(n_pieces);
+    for (tr_piece_index_t i = 0; i < n_pieces; ++i)
+    {
+        if (!peer_info.clientCanRequestPiece(i))
+        {
+            continue;
+        }
+
+        size_t const n_missing = peer_info.countMissingBlocks(i);
+        if (n_missing == 0)
+        {
+            continue;
+        }
+
+        wanted_pieces.emplace_back(i, n_missing);
+    }
+
+    // transform them into candidates
+    auto const n = std::size(wanted_pieces);
+    auto saltbuf = std::vector<char>(n);
+    tr_rand_buffer(std::data(saltbuf), n);
+    auto candidates = std::vector<Candidate>{};
+    candidates.reserve(n);
+    for (size_t i = 0; i < n; ++i)
+    {
+        auto const [piece, n_missing] = wanted_pieces[i];
+        candidates.emplace_back(piece, n_missing, peer_info.priority(piece), saltbuf[i]);
+    }
+
+    return candidates;
+}
+
+static std::vector<tr_block_range_t> makeRanges(tr_block_index_t const* sorted_blocks, size_t n_blocks)
+{
+    if (n_blocks == 0)
+    {
+        // std::cout << __FILE__ << ':' << __LINE__ << " makeRanges returning empty vector" << std::endl;
+        return {};
+    }
+
+#if 0
```

Comment:
Leftovers?

Reply:
Yeah the `#if 0` are for testing. I'm still playing around with it.  Guess I coulda done TDD instead :slightly_smiling_face: